pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/RealtimeToOfflineSegmentsTaskGenerator.java

```diff
@@ -0,0 +1,238 @@
+
+      // Get the bucket size and buffer
+      String bucketTimeStr =
+          taskConfigs.getOrDefault(RealtimeToOfflineSegmentsTask.BUCKET_TIME_PERIOD_KEY, DEFAULT_BUCKET_PERIOD);
+      String bufferTimeStr =
+          taskConfigs.getOrDefault(RealtimeToOfflineSegmentsTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_PERIOD);
+      long bucketMillis = TimeUtils.convertPeriodToMillis(bucketTimeStr);
+      long bufferMillis = TimeUtils.convertPeriodToMillis(bufferTimeStr);
+
+      // Fetch RealtimeToOfflineSegmentsTaskMetadata ZNode for reading watermark
+      RealtimeToOfflineSegmentsTaskMetadata realtimeToOfflineSegmentsTaskMetadata =
+          _clusterInfoProvider.getMinionRealtimeToOfflineSegmentsTaskMetadata(realtimeTableName);
+
+      if (realtimeToOfflineSegmentsTaskMetadata == null) {
+        // No ZNode exists. Cold-start.
+        long watermarkMillis;
+
+        String startTimeStr = taskConfigs.get(RealtimeToOfflineSegmentsTask.START_TIME_MILLIS_KEY);
+        if (startTimeStr != null) {
+          // Use startTime config if provided in taskConfigs
+          watermarkMillis = Long.parseLong(startTimeStr);
+        } else {
+          // Find the smallest time from all segments
+          RealtimeSegmentZKMetadata minSegmentZkMetadata = null;
+          for (LLCRealtimeSegmentZKMetadata realtimeSegmentZKMetadata : completedSegmentsMetadataList) {
+            if (minSegmentZkMetadata == null || realtimeSegmentZKMetadata.getStartTime() < minSegmentZkMetadata
+                .getStartTime()) {
+              minSegmentZkMetadata = realtimeSegmentZKMetadata;
+            }
+          }
+          Preconditions.checkState(minSegmentZkMetadata != null);
+
+          // Convert the segment minTime to millis
+          long minSegmentStartTimeMillis =
+              minSegmentZkMetadata.getTimeUnit().toMillis(minSegmentZkMetadata.getStartTime());
+
+          // Round off according to the bucket. This ensures we align the offline segments to proper time boundaries
+          // For example, if start time millis is 20200813T12:34:59, we want to create the first segment for window [20200813, 20200814)
+          watermarkMillis = (minSegmentStartTimeMillis / bucketMillis) * bucketMillis;
+        }
+
+        // Create RealtimeToOfflineSegmentsTaskMetadata ZNode using watermark calculated above
+        realtimeToOfflineSegmentsTaskMetadata =
+            new RealtimeToOfflineSegmentsTaskMetadata(realtimeTableName, watermarkMillis);
+        _clusterInfoProvider.setRealtimeToOfflineSegmentsTaskMetadata(realtimeToOfflineSegmentsTaskMetadata);
+      }
+
+      // WindowStart = watermark. WindowEnd = windowStart + bucket.
+      long windowStartMillis = realtimeToOfflineSegmentsTaskMetadata.getWatermarkMillis();
+      long windowEndMillis = windowStartMillis + bucketMillis;
+
+      // Check that execution window is older than bufferTime
+      if (windowEndMillis > System.currentTimeMillis() - bufferMillis) {
```

Comment:
if you use a method to get the current time, you can test the logic in this class by manipulating time to anything.